### PR TITLE
[Resolved] Open resource link externally button

### PR DESCRIPTION
**Pull request by Mohit-Joshi-dev**
[https://github.com/guyluz11/infinite_horizons/issues/109](https://github.com/guyluz11/infinite_horizons/issues/109)

Added URL launcher on resources page.

LaunchURL

### CODE REVIEW

**Comment by Mohit-Joshi-dev** [comment]
@guyluz11

File: lib/presentation/pages/tip_resources_page.dart [hunk 1]
```diff
@@ -31,9 +34,17 @@ class TipResourcePage extends StatelessWidget {
...
         url: url,
       ),
     );
   }
+
+  Future<void> openUrl(String url) async {
+    if (!await launchUrl(Uri.parse(url))) {
+      throw Exception('Could not launch $url');
+    }
+  }
```
Comment (guyluz11) [inline lib/presentation/pages/tip_resources_page.dart:49]: Can you create a controller that calls launchUrl methods?.
Follow the existing controllers as an example.

File: lib/presentation/pages/tip_resources_page.dart [hunk 2]
```diff
@@ -16,7 +17,9 @@ class TipResourcePage extends StatelessWidget {
         title: "resource",
         margin: false,
         topBarType: TopBarType.back,
+        topBarRightOnTap: () => openUrl(url),
         expendChild: false,
+        topBarRightIcon: Icons.open_in_new,
         child: PdfViewerMolecule(
           url: url,
         ),
```
Comment (guyluz11) [inline lib/presentation/pages/tip_resources_page.dart:25]: You are missing open url button for the YoutubePlayerPage

**Review by guyluz11 — CHANGES_REQUESTED** [review]

**Comment by Mohit-Joshi-dev** [comment]
@guyluz11 I have completed requested changes

File: lib/infrastructure/url_launcher_repository.dart [hunk 3]
```diff
@@ -0,0 +1,10 @@
+part of 'package:infinite_horizons/domain/controllers/url_launcher_controller.dart';
+
+class _URLLauncherRepository extends URLLauncherController {
```
Comment (guyluz11) [inline lib/infrastructure/url_launcher_repository.dart:3]: Change name to _UrlLauncherRepository

File: lib/domain/controllers/url_launcher_controller.dart [hunk 4]
```diff
@@ -0,0 +1,12 @@
+import 'package:url_launcher/url_launcher.dart';
+
+part 'package:infinite_horizons/infrastructure/url_launcher_repository.dart';
+
+abstract class URLLauncherController {
```
Comment (guyluz11) [inline lib/domain/controllers/url_launcher_controller.dart:5]: Change name to UrlLauncherRepository

File: lib/infrastructure/url_launcher_repository.dart [hunk 5]
```diff
@@ -0,0 +1,10 @@
...
+  @override
+  Future<void> openUrl(String url) async {
+    if (!await launchUrl(Uri.parse(url))) {
+      throw Exception('Could not launch $url');
```
Comment (guyluz11) [inline lib/infrastructure/url_launcher_repository.dart:7]: Let's return bool instead

File: lib/presentation/pages/tip_resources_page.dart [hunk 6]
```diff
@@ -1,4 +1,5 @@
 import 'package:flutter/material.dart';
+import 'package:infinite_horizons/domain/controllers/url_launcher_controller.dart';
```
Comment (guyluz11) [inline lib/presentation/pages/tip_resources_page.dart:2]: Please export `url_launcher_controller.dart';` from `controllers.dart` and import controllers.dart instead instead

**Review by guyluz11 — CHANGES_REQUESTED** [review]

**Review by guyluz11 — APPROVED** [review]

**Pull request merged**